src/linear_algebra/unitary_group.lean

```diff
@@ -0,0 +1,176 @@
+/-
+Copyright (c) 2021 Shing Tak Lam. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Shing Tak Lam
+-/
+import linear_algebra.matrix
+import linear_algebra.nonsingular_inverse
+import data.complex.basic
+
+/-!
+# The Unitrary Group
+
+This file defines elements of the unitary group `unitary_group n α`, where `α` is a `star_ring`.
+This consists of all `n` by `n` matrices with entries in `α` such that the star-transpose is its
+inverse. In addition, we define the group structure on `unitary_group n α`, and the embedding into
+the general linear group `general_linear_group α (n → α)`.
+
+We also define the orthogonal group `orthogonal_group n β`, where `β` is a `comm_ring`.
+
+## Main Definitions
+
+ * `matrix.unitary_group` is the type of matrices where the star-transpose is the inverse
+ * `matrix.unitary_group.group` is the group structure (under multiplication)
+ * `matrix.unitary_group.embedding_GL` is the embedding `unitary_group n α → GLₙ(α)`
+ * `matrix.orthogonal_group` is the type of matrices where the transpose is the inverse
+
+## References
+
+ * https://en.wikipedia.org/wiki/Unitary_group
+
+## Tags
+
+matrix group, group, unitary group, orthogonal group
+
+-/
+
+namespace matrix
+universes u v
+open linear_map
+open_locale matrix
+
+
+section
+
+variables (n : Type u) [fintype n] [decidable_eq n]
+variables (α : Type v) [comm_ring α] [star_ring α]
+
+/--
+  `unitary_group n` is the group of `n` by `n` complex matrices where the conjugate transpose is
+  the inverse.
+-/
+def unitary_group : Type* := {A : matrix n n α // star A ⬝ A = 1}
```

Comment:
I still wonder if it would be easier to start by defining
```
def unitary_submonoid : submonoid (matrix n n α) := sorry

@[derive [monoid, has_coe_to_fun]]
def unitary_group : Type := unitary_submonoid
```
which would give you the one, mul, and monoid structure for free

Reply:
Yes. I've tried changing it to a submonoid based definition, and it doesn't seem to be as nice? Since in a lot of definitions I need the fact that `star A * A = 1`, and if it is defined through a submonoid then there's an extra layer or two of abstraction.

Right now when I need that fact, it's just `subtype.property`, or `A.2`. But with the submonoid def, `A.2` gives me `A.val ∈ unitary_submonoid n α`, which is defeq to `star A.1 * A.1 = 1`, but `rw` doesn't work so I'll need a `have` or a `show` or another lemma to get it to be the right form.

Reply:
I like the idea of starting with the submonoid definition, too, though I don't want to push it if it's unworkable.  Can you do the submonoid definition, and just make your first lemma the fact that `star A * A = 1`?

Reply:
Yup, that works.